core/src/main/java/org/infinispan/marshall/core/GlobalMarshaller.java

```diff
@@ -370,8 +414,189 @@ private void writeNonNullableObject(Object obj, BytesObjectOutput out) throws IO
       }
    }
 
+   private void writeArray(Class<?> clazz, Object array, BytesObjectOutput out) throws IOException {
+      out.writeByte(ID_ARRAY);
+      Class<?> componentType = clazz.getComponentType();
+
+      int length = Array.getLength(array);
+      boolean singleType = true;
+      Class<?> elementType = null;
+      int flags;
+      if (length == 0) {
+         flags = FLAG_ARRAY_EMPTY;
+      } else {
+         if (length <= Primitives.SMALL_ARRAY_MAX) {
+            flags = FLAG_ARRAY_SMALL;
+         } else if (length <= Primitives.MEDIUM_ARRAY_MAX) {
+            flags = FLAG_ARRAY_MEDIUM;
+         } else {
+            flags = FLAG_ARRAY_LARGE;
+         }
+         Object firstElement = Array.get(array, 0);
+         if (firstElement != null) {
+            elementType = firstElement.getClass();
+         }
+         for (int i = 1; i < length; ++i) {
```

Comment:
I would assume that just checking all the elements is pretty cheap, compared to everything else that's going on there. However, assuming that and backing off is an interesting strategy, I'll think about that. Not sure if that's worth the effort, though.

Ad creating the array 'ourselves' - this is a generic marshalling code, I don't see how would you pass that hint to marshaller from the place where you create the array.